Backends/CLX/clipboard.lisp

```diff
@@ -0,0 +1,249 @@
+            (call-selection-translator (selection-object-content result)
+                                       (selection-object-type result)
+                                       to-type
+                                       table)))))))
+
+(macrolet ((frob (selection)
+             `(defmethod request-selection ((port clx-selection-mixin)
+                                            (selection (eql ,selection))
+                                            (requested-object selection-object))
+                (when (climi::stored-object port selection)
+                  (return-from request-selection (call-next-method)))
+                (clx-request-selection port selection requested-object))))
+  (frob :primary)
+  (frob :clipboard))
+
+;;; define basic translators
+(define-presentation-type clx-string ())
+(alexandria:define-constant +clx-string-names+
+    '(:utf8_string :text :string :|text/plain;charset=utf-8| :|text/plain|)
+  :test #'equal)
+(macrolet ((frob ()
+             `(progn ,@(mapcar (lambda (ptype)
+                                 `(define-presentation-type ,ptype () :inherit-from 'clx-string))
+                               +clx-string-names+))))
+  ;; selection -> string
+  (clime:define-selection-translator clx-string->lisp-string
+      (clx-string string global-command-table)
+      (object)
+    (babel:octets-to-string (coerce object '(vector (unsigned-byte 8))) :encoding :utf-8))
+  (frob))
+
+(defun process-selection-request (port window sheet target property requestor selection time
+                                  &aux object from-type table)
+  (declare (ignorable window sheet))
+  (when (setf object (climi::stored-object port selection))
+    (setf from-type (selection-object-type object)
+          table (climi::get-object-table object)))
+  (unless (and property object)
+    ;; we can't set NIL property (this may happen with bogus peers) -- jd 2019-06-04
+    (xlib:send-event requestor :selection-notify nil
+                     :event-window requestor :selection selection :target target
+                     :property property :time time)
+    (setf (clipboard-outstanding-request port) nil)
+    (return-from process-selection-request))
+  (cond
+    ;; We treat TARGETS specially: we look for all selection translators from the type and return
+    ;; them. If the target type is a keyword, we return it as is. Strings are treated in a similar
+    ;; manner.
+    ((eql target :targets)
+     (let* ((to-type (loop
+                        for tr in (find-selection-translators from-type t table)
+                        for type = (climi::to-type tr)
+                        when (or (keywordp type) (eql type 'string)) collect type))
+            (targets (mapcar (alexandria:curry #'xlib:intern-atom (xlib:window-display window))
+                             (if (or (keywordp from-type) (eql from-type 'string))
+                                 (list* :targets from-type to-type)
```

Comment:
There is a problem here. `:STRING` is only a valid type if the string itself is encodable using plain ASCII (or possibly ISO8859-1, the documentation I've read isn't clear on this topic). Also, we always want to include `:UTF8_STRING` and `:|text/plain;charset=utf8|` here as well.

Also, I wasn't aware that `XLIB:INTERN-ATOM` accepted non-keyword symbols. But, it seems to do that.

Reply:
fixed